sled-agent/src/sled_agent.rs

```diff
@@ -167,10 +168,19 @@ impl SledAgent {
         // re-establish contact (i.e., if the Sled Agent crashed, but we wanted
         // to leave the running Zones intact).
         let zones = Zones::get()?;
-        for z in zones {
-            warn!(log, "Deleting existing zone"; "zone_name" => z.name());
-            Zones::halt_and_remove_logged(&log, z.name())?;
-        }
+        let logs = vec![log.clone(); zones.len()];
```

Comment:
Interesting! I'd have expected that `Zones::halt_and_removed_logged` takes an exclusive reference, this could be shared. Is the `move` on that closure what prevents that? I'm just curious, I think this is fine as-is.

Reply:
With this code (which is basically the existing version in the PR, but no cloned logs. Note, both accesses of `log` use `&log`!):

```rust
        let zones = Zones::get()?;                                                                                   
        stream::iter(zones)                                                                                          
            .map(Ok::<_, crate::zone::AdmError>)          
            .try_for_each_concurrent(                                                                                
                None,                                                                                                
                |zone| async {                                                                                       
                    tokio::task::spawn_blocking(move || { 
                        warn!(&log, "Deleting existing zone"; "zone_name" => zone.name());                            
                        Zones::halt_and_remove_logged(&log, zone.name())                                             
                    }).await.unwrap()                     
                }
            ).await?;
```

I see the following error:

```
error[E0507]: cannot move out of `log`, a captured variable in an `FnMut` closure
   --> sled-agent/src/sled_agent.rs:177:30
    |
127 |            let log = log.new(o!(
    |                --- captured outer variable
...
177 |                    |zone| async {
    |  __________________-____________^
    | | _________________|
    | ||
178 | ||                     tokio::task::spawn_blocking(move || {
179 | ||                         warn!(&log, "Deleting existing zone"; "zone_name" => zone.name());
    | ||                                ---
    | ||                                |
    | ||                                variable moved due to use in generator
    | ||                                move occurs because `log` has type `Logger<Arc<dyn SendSyncRefUnwindSafeDrain<Err = slog::private::NeverStruct, Ok = ()>>>`, which does not implement the `Copy` trait
180 | ||                         Zones::halt_and_remove_logged(&log, zone.name())
181 | ||                     }).await.unwrap()
182 | ||                 }
    | ||                 ^
    | ||_________________|
    | |__________________captured by this `FnMut` closure
    |                    move out of `log` occurs here

```

Even when I tried to ensure "log" was actually a "reference to the log", it complained it was not static.

Reply:
[spawn_blocking](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html) has a `'static` bound on the closure it accepts, so it's not possible to avoid moving an owned `log` in. That said, I think you could avoid creating the temporary `Vec` of logs via `std::iter::repeat()`; something like

```rust
        stream::iter(zones)
            .zip(stream::iter(std::iter::repeat(log.clone())))
            // rest unchanged
```

Reply:
oooo, nice suggestion @jgallagher  - updated!